030-extract-core-resources/proposal.md

```diff
@@ -0,0 +1,79 @@
+# Summary
+
+Concourse currently ships with the following "core" resource types:
+
+* [`bosh-io-release`](https://github.com/concourse/bosh-io-release-resource)
+* [`bosh-io-stemcell`](https://github.com/concourse/bosh-io-stemcell-resource)
+* [`cf`](https://github.com/concourse/cf-resource)
+* [`docker-image`](https://github.com/concourse/docker-image-resource)
+* [`git`](https://github.com/concourse/git-resource)
+* [`github-release`](https://github.com/concourse/github-release-resource)
+* [`hg`](https://github.com/concourse/hg-resource)
+* [`mock`](https://github.com/concourse/mock-resource)
+* [`pool`](https://github.com/concourse/pool-resource)
+* [`registry-image`](https://github.com/concourse/registry-image-resource)
+* [`s3`](https://github.com/concourse/s3-resource)
+* [`semver`](https://github.com/concourse/semver-resource)
+* [`time`](https://github.com/concourse/time-resource)
+* [`tracker`](https://github.com/concourse/tracker-resource)
+
+This proposal is to remove them all from the `.tgz` distribution, leaving only the `registry-image` resource, as it will be necessary for fetching other resource types.
+
+# Motivation
+
+1. Packaging them all up with Concourse inflates the download size, especially as we standardize on `ubuntu`  as the base image instead of `alpine`.
+
+1. Having commonly-used resource types bundled into the core set makes them feel "special/blessed" and may result in a delayed introduction to/understanding of `resource_types:` for new users.
+
+1. A few of these resource types are awfully specific and not likely to be used by the majority of Concourse users:
+
+    * `bosh-io-release`
+    * `bosh-io-stemcell`
+    * `cf`
+    * `tracker`
+
+# Proposal
+
+Removing them is the easy part. The harder part is doing this without requiring everyone to update their pipelines.
+
+We could do this by having an operator-specified config of resource types to automatically provide to each pipeline:
+
+```yaml
+---
+resource_types:
+- name: git
+  type: registry-image
+  source:
+    repository: concourse/git-resource
+    tag: 1
+- name: time
+  type: registry-image
+  source:
+    repository: concourse/time-resource
+    tag: 1
+```
+
+This would be passed to `concourse web` on startup like so:
+
+```sh
+concourse web ... --core-resource-types resource-types.yml
+```
+
+This would have the effect of being merged into each pipeline's `resource_types:` (with the pipeline taking precedence).
+
+Operators should be careful to provide a sane `tag:` value so that newly shipped versions do not break all pipelines. At the very least a major version should be specified. This is up to the operator's discretion, however, and will not be enforced.
+
+# Open Questions
+
+* This will result in a lot more polling against the Docker registry. Should we have a longer default check interval for resource types?
```

Comment:
Maybe even disable checking for new version? To keep the exact behaviour as now, where included resource types are only updated when updating Concourse. Otherwise this introduces a kind of silent updates that can't be blocked

Reply:
If these core resources would be pinned to specific image digest (ie the `sha256:…` value) then there would be no need to check registry again, right?

But looks like current `registry-image` resource doesn't support digest-based pinning yet.

Reply:
Currently resources can use [`version`](https://concourse-ci.org/resources.html#resource-version) when specifying an exact version of resource to use. I could see this being added to resource types.

Reply:
As of right now, we do not check base resource types. We could retain this behaviour to avoid the extra checks.
 
Currently, we only check resource types which are defined as part of a pipeline. There is no such thing as a resource type that does not exist in a pipeline. However if we modeled these core resource types as regular resource types, no checks would occur because they are not defined in a pipeline. 

When fetching these resources, we could just fetch `latest` by default or the `version` specified in the yaml as @jtarchie suggested. If you want a new version of the base resource type, you have to change the `resource-types.yml` and restart the ATC.

Reply:
One thing to think about:

Resource type versions are taken into account when a resource's version history is composed, to be shown in UI. Currently (at least as of the v5.x line - sorry I'm not exactly sure when this change was introduced) when custom resource type's image is updated, versions fetched with an older version of the resource type aren't shown by default in the resource UI. If the fetching mechanism is changed across the board for all resource types, then users are going to start seeing this edge case a lot more frequently. Here's more context on the issue that I saw: https://github.com/concourse/concourse/issues/4893

Reply:
I'd definitely vote for erring on the side of a longer check interval. Like even once a day. But maybe this can be configurable in the `resource-types.yml` that operators are feeding in for pipeline-wide config.

Probably quite a few Concourse users are routing outgoing traffic through load balancers or NAT boxes or stuff that will make it look like all traffic is coming from the same IP, and we're starting to see rate limiting and throttling from Docker Hub. This is not scientific evidence but in this issue thread, it looks like lots of users of other CI/CD tools have also experienced similar: https://github.com/docker/hub-feedback/issues/1907